File: src/code_indexing_mcp/acceptance.py
"""Pure correctness metrics shared by accelerator promotion gates."""

from __future__ import annotations

import math
from collections.abc import Sequence
from typing import Any

import numpy as np
from numpy.typing import NDArray

FloatMatrix = NDArray[np.float32]


def _matrix(values: NDArray[Any], *, name: str) -> FloatMatrix:
    matrix = np.asarray(values, dtype=np.float32)
    if matrix.ndim != 2 or not matrix.shape[0] or not matrix.shape[1]:
        raise ValueError(f"{name} must be a non-empty two-dimensional matrix")
    if not np.all(np.isfinite(matrix)):
        raise ValueError(f"{name} contains non-finite values")
    return matrix


def _normalized(values: NDArray[Any], *, name: str) -> FloatMatrix:
    matrix = _matrix(values, name=name)
    norms = np.linalg.norm(matrix, axis=1, keepdims=True)
    if np.any(norms <= np.float32(1e-12)):
        raise ValueError(f"{name} contains a zero-length row")
    return np.asarray(matrix / norms, dtype=np.float32)


def cosine_rows(reference: NDArray[Any], candidate: NDArray[Any]) -> FloatMatrix:
    """Return cosine similarity for each corresponding pair of matrix rows."""

    left = _normalized(reference, name="reference")
    right = _normalized(candidate, name="candidate")
    if left.shape != right.shape:
        raise ValueError(f"reference and candidate shapes differ: {left.shape} != {right.shape}")
    return np.asarray(np.sum(left * right, axis=1), dtype=np.float32)


def top_k_overlap(
    queries: NDArray[Any],
    reference: NDArray[Any],
    candidate: NDArray[Any],
    *,
    k: int,
) -> float:
    """Return mean top-k result overlap for two candidate vector matrices."""

    query_rows = _normalized(queries, name="queries")
    reference_rows = _normalized(reference, name="reference")
    candidate_rows = _normalized(candidate, name="candidate")
    if reference_rows.shape != candidate_rows.shape:
        raise ValueError(
            f"reference and candidate shapes differ: "
            f"{reference_rows.shape} != {candidate_rows.shape}"
        )
    if query_rows.shape[1] != reference_rows.shape[1]:
        raise ValueError(
            f"query dimension {query_rows.shape[1]} does not match "
            f"candidate dimension {reference_rows.shape[1]}"
        )
    if not 1 <= k <= reference_rows.shape[0]:
        raise ValueError(f"k must be from 1 to {reference_rows.shape[0]}")

    reference_order = np.argsort(
        -(query_rows @ reference_rows.T),
        axis=1,
        kind="stable",
    )[:, :k]
    candidate_order = np.argsort(
        -(query_rows @ candidate_rows.T),
        axis=1,
        kind="stable",
    )[:, :k]
    overlap = sum(
        len(set(left.tolist()).intersection(right.tolist()))
        for left, right in zip(reference_order, candidate_order, strict=True)
    )
    return float(overlap / (query_rows.shape[0] * k))


def top_k_rank_correlation(
    reference_orders: Sequence[Sequence[Any]],
    candidate_orders: Sequence[Sequence[Any]],
) -> float:
    """Return the mean Kendall tau-b between paired top-k id rankings.

    Both rankings are top-k windows over one result set of unique ids, so an
    id present in only one window counts as tied just past that window's end
    (position ``k``) in the other: losing a result outright is a tie among the
    lost rather than an invented rank. 1.0 is an identical ranking and -1.0 a
    fully reversed one over the same ids.
    """

    if len(reference_orders) != len(candidate_orders):
        raise ValueError(
            f"ranking counts differ: {len(reference_orders)} != {len(candidate_orders)}"
        )
    if not reference_orders:
        raise ValueError("at least one pair of rankings is required")
    scores: list[float] = []
    for reference_order, candidate_order in zip(reference_orders, candidate_orders, strict=True):
        if not reference_order or not candidate_order:
            raise ValueError("a top-k ranking must not be empty")
        window = max(len(reference_order), len(candidate_order))
        union = list(dict.fromkeys([*reference_order, *candidate_order]))
        reference_position = {item: index for index, item in enumerate(reference_order)}
        candidate_position = {item: index for index, item in enumerate(candidate_order)}
        reference_ranks = np.asarray(
            [reference_position.get(item, window) for item in union], dtype=np.int64
        )
        candidate_ranks = np.asarray(
            [candidate_position.get(item, window) for item in union], dtype=np.int64
        )
        left, right = np.triu_indices(len(union), k=1)
        reference_sign = np.sign(reference_ranks[left] - reference_ranks[right])
        candidate_sign = np.sign(candidate_ranks[left] - candidate_ranks[right])
        concordant = int(np.sum((reference_sign != 0) & (reference_sign == candidate_sign)))
        discordant = int(np.sum((reference_sign != 0) & (reference_sign == -candidate_sign)))
        tied_reference = int(np.sum((reference_sign == 0) & (candidate_sign != 0)))
        tied_candidate = int(np.sum((candidate_sign == 0) & (reference_sign != 0)))
        denominator = math.sqrt(
            float(
                (concordant + discordant + tied_reference)
                * (concordant + discordant + tied_candidate)
            )
        )
        # Both windows hold exactly the same single id: identical rankings.
        scores.append(1.0 if denominator == 0.0 else (concordant - discordant) / denominator)
    return float(sum(scores) / len(scores))

File: ts/packages/server/src/acceptance.ts
/**
 * Pure correctness metrics shared by accelerator promotion gates.
 *
 * Written over plain arrays rather than a linear-algebra dependency: the
 * matrices here are a probe corpus, not a workload, and the whole module is on
 * no hot path. The Python original computes in float32 because that is what
 * numpy narrows an embedding matrix to; this one stays in float64, which can
 * only move a cosine by around 1e-7 -- five orders of magnitude below the
 * loosest gate any caller applies, and already confirmed harmless by the S3
 * embedding-parity spike.
 */

/** A non-empty, rectangular matrix of finite numbers. */
export type Matrix = readonly (readonly number[])[];

function validated(matrix: Matrix, name: string): Matrix {
  const width = matrix[0]?.length ?? 0;
  if (matrix.length === 0 || width === 0) {
    throw new Error(`${name} must be a non-empty two-dimensional matrix`);
  }
  for (const row of matrix) {
    if (row.length !== width) {
      throw new Error(`${name} must be a non-empty two-dimensional matrix`);
    }
    for (const value of row) {
      if (!Number.isFinite(value)) throw new Error(`${name} contains non-finite values`);
    }
  }
  return matrix;
}

/** L2-normalize each row, rejecting the zero-length rows a cosine cannot score. */
function normalized(matrix: Matrix, name: string): number[][] {
  return validated(matrix, name).map((row) => {
    const norm = Math.sqrt(row.reduce((total, value) => total + value * value, 0));
    if (norm <= 1e-12) throw new Error(`${name} contains a zero-length row`);
    return row.map((value) => value / norm);
  });
}

function shape(matrix: Matrix): string {
  return `(${matrix.length}, ${matrix[0]?.length ?? 0})`;
}

function dot(left: readonly number[], right: readonly number[]): number {
  return left.reduce((total, value, index) => total + value * (right[index] ?? 0), 0);
}

/** Return cosine similarity for each corresponding pair of matrix rows. */
export function cosineRows(reference: Matrix, candidate: Matrix): number[] {
  const left = normalized(reference, "reference");
  const right = normalized(candidate, "candidate");
  if (left.length !== right.length || (left[0]?.length ?? 0) !== (right[0]?.length ?? 0)) {
    throw new Error(`reference and candidate shapes differ: ${shape(left)} != ${shape(right)}`);
  }
  return left.map((row, index) => dot(row, right[index] ?? []));
}

/** Indices of the k highest scores, ties broken by ascending index. */
function topKIndices(scores: readonly number[], k: number): number[] {
  return scores
    .map((score, index) => ({ score, index }))
    .sort((left, right) => right.score - left.score || left.index - right.index)
    .slice(0, k)
    .map((entry) => entry.index);
}

/** Return mean top-k result overlap for two candidate vector matrices. */
export function topKOverlap(
  queries: Matrix,
  reference: Matrix,
  candidate: Matrix,
  { k }: { k: number },
): number {
  const queryRows = normalized(queries, "queries");
  const referenceRows = normalized(reference, "reference");
  const candidateRows = normalized(candidate, "candidate");
  if (
    referenceRows.length !== candidateRows.length ||
    (referenceRows[0]?.length ?? 0) !== (candidateRows[0]?.length ?? 0)
  ) {
    throw new Error(
      `reference and candidate shapes differ: ` +
        `${shape(referenceRows)} != ${shape(candidateRows)}`,
    );
  }
  if ((queryRows[0]?.length ?? 0) !== (referenceRows[0]?.length ?? 0)) {
    throw new Error(
      `query dimension ${queryRows[0]?.length ?? 0} does not match ` +
        `candidate dimension ${referenceRows[0]?.length ?? 0}`,
    );
  }
  if (!(k >= 1 && k <= referenceRows.length)) {
    throw new Error(`k must be from 1 to ${referenceRows.length}`);
  }

  let overlap = 0;
  for (const query of queryRows) {
    const referenceOrder = new Set(
      topKIndices(
        referenceRows.map((row) => dot(query, row)),
        k,
      ),
    );
    for (const index of topKIndices(
      candidateRows.map((row) => dot(query, row)),
      k,
    )) {
      if (referenceOrder.has(index)) overlap += 1;
    }
  }
  return overlap / (queryRows.length * k);
}

/**
 * Return the mean Kendall tau-b between paired top-k id rankings.
 *
 * Both rankings are top-k windows over one result set of unique ids, so an id
 * present in only one window counts as tied just past that window's end
 * (position `k`) in the other: losing a result outright is a tie among the lost
 * rather than an invented rank. 1.0 is an identical ranking and -1.0 a fully
 * reversed one over the same ids.
 */
export function topKRankCorrelation<T>(
  referenceOrders: readonly (readonly T[])[],
  candidateOrders: readonly (readonly T[])[],
): number {
  if (referenceOrders.length !== candidateOrders.length) {
    throw new Error(
      `ranking counts differ: ${referenceOrders.length} != ${candidateOrders.length}`,
    );
  }
  if (referenceOrders.length === 0) {
    throw new Error("at least one pair of rankings is required");
  }
  let total = 0;
  for (const [index, referenceOrder] of referenceOrders.entries()) {
    const candidateOrder = candidateOrders[index] ?? [];
    if (referenceOrder.length === 0 || candidateOrder.length === 0) {
      throw new Error("a top-k ranking must not be empty");
    }
    total += kendallTauB(referenceOrder, candidateOrder);
  }
  return total / referenceOrders.length;
}

function kendallTauB<T>(referenceOrder: readonly T[], candidateOrder: readonly T[]): number {
  const window = Math.max(referenceOrder.length, candidateOrder.length);
  const union = [...new Set([...referenceOrder, ...candidateOrder])];
  const rank = (order: readonly T[]): number[] => {
    const positions = new Map(order.map((item, position) => [item, position]));
    return union.map((item) => positions.get(item) ?? window);
  };
  const referenceRanks = rank(referenceOrder);
  const candidateRanks = rank(candidateOrder);

  let concordant = 0;
  let discordant = 0;
  let tiedReference = 0;
  let tiedCandidate = 0;
  for (let left = 0; left < union.length; left += 1) {
    for (let right = left + 1; right < union.length; right += 1) {
      const referenceSign = Math.sign(
        (referenceRanks[left] ?? window) - (referenceRanks[right] ?? window),
      );
      const candidateSign = Math.sign(
        (candidateRanks[left] ?? window) - (candidateRanks[right] ?? window),
      );
      if (referenceSign !== 0 && referenceSign === candidateSign) concordant += 1;
      if (referenceSign !== 0 && referenceSign === -candidateSign) discordant += 1;
      if (referenceSign === 0 && candidateSign !== 0) tiedReference += 1;
      if (candidateSign === 0 && referenceSign !== 0) tiedCandidate += 1;
    }
  }
  const denominator = Math.sqrt(
    (concordant + discordant + tiedReference) * (concordant + discordant + tiedCandidate),
  );
  // Both windows hold exactly the same single id: identical rankings.
  return denominator === 0 ? 1 : (concordant - discordant) / denominator;
}
